geo/geo.go

```diff
@@ -0,0 +1,246 @@
+		return false
+	}
+	for _, loc := range f.Allow {
+		if len(loc.CountryCode) > 0 || len(loc.Country) > 0 || len(loc.City) > 0 {
+			return true
+		}
+	}
+	for _, loc := range f.Disallow {
+		if len(loc.CountryCode) > 0 || len(loc.Country) > 0 || len(loc.City) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (f *IPFilter) AllowIP(ip string) (bool, error) {
+	if f.Empty() {
+		return true, nil
+	}
+
+	var loc *Location
+	if f.NeedGeoInfo() {
+		loc = f.GetLocation(ip)
+	} else {
+		loc = &Location{IP: ip}
+	}
+
+	return f.AllowLocation(loc), nil
+}
+
+func (f *IPFilter) GetLocation(ip string) *Location {
+	for _, p := range f.Providers {
+		if p.Ready() {
+			loc := getLocationFromProvider(ip, p.(GeoProvider))
+			if !loc.Empty() {
+				return &loc
+			}
+		}
+	}
+	return &Location{CountryCode: "UNKNOWN", IP: ip}
+}
+
+func (f *IPFilter) AllowLocation(loc *Location) bool {
+	if loc.Empty() {
+		return true
+	}
+
+	for _, l := range f.Disallow {
+		if l.Match(loc) {
+			return false
+		}
+	}
+
+	empty := true
+	for _, l := range f.Allow {
+		if l.Match(loc) {
+			return true
+		}
+		if !l.Empty() {
+			empty = false
+		}
+	}
+
+	return empty
+}
+
+func getLocationFromProvider(ip string, p GeoProvider) Location {
+	loc, err := p.GetLocation(ip)
+	if err != nil {
+		log.Println(err)
+	}
+	return *loc
+}
+
+func GetJsonFile(url, filename string) ([]byte, error) {
+	if s, err := os.Stat(filename); os.IsNotExist(err) || s.Size() == 0 {
+		f, err := os.Create(filename)
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+
+		client := http.Client{
+			Timeout: 5 * time.Second,
+		}
+		resp, err := client.Get(url)
+		if err != nil {
+			os.Remove(filename)
+			return nil, err
+		}
+		defer resp.Body.Close()
+		b, err := ioutil.ReadAll(resp.Body)
+		if !json.Valid(b) {
+			os.Remove(filename)
+			return nil, errors.New("invalid json")
+		}
+		_, err = f.Write(b)
+		if err != nil {
+			os.Remove(filename)
+			return nil, err
+		}
+	}
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	b, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func (f *IPFilter) AddProvider() {
+	if f.DownloadDB {
+		aws := NewAWSProvider(f.DBPath)
+		err := aws.MaybeUpdate()
+		if err != nil {
+			log.Print(err)
+		} else {
+			f.Providers = append(f.Providers, aws)
+		}
+
+		gcp := NewGCPProvider(f.DBPath)
+		err = gcp.MaybeUpdate()
+		if err != nil {
+			log.Print(err)
+		} else {
+			f.Providers = append(f.Providers, gcp)
+		}
+
+		mm := NewMaxMindProvider(f.DBPath)
+		err = mm.MaybeUpdate()
+		if err != nil {
+			log.Print(err)
+		} else {
+			f.Providers = append(f.Providers, mm)
+		}
+	}
+
+	ip2c := NewIP2CProvider()
+	ip2c.MaybeUpdate()
+	f.Providers = append(f.Providers, ip2c)
+}
+
+func (f *IPFilter) UpdateDataFile() {
+	for {
+		time.Sleep(10 * time.Second)
+		for idx, p := range f.Providers {
+			if len(p.FileName()) == 0 {
+				continue
+			}
+			if !p.NeedUpdate() {
```

Comment:
Is this needed? `NeedUpdate()` is called inside `MaybeUpdate()` protected by lock, but this call here is not protected by lock